### PR TITLE
#task8

### PR DESCRIPTION
**Pull request by YauheniRaby**

### CODE REVIEW

File: Solution1/BusinessLayer/Services/WeatherService.cs [hunk 1]
```diff
@@ -89,7 +97,12 @@ public async Task<ForecastWeatherDTO> GetForecastByCityNameAsync(string cityName
                     else
                     {
                         weatherResponseDTO.ErrorMessage = validationResult.Errors.FirstOrDefault().ErrorMessage;
```
Comment (elizavetapavliv) [inline Solution1/BusinessLayer/Services/WeatherService.cs:99]: now you can specify failed status explicitly, cause it's not so obviously that Fail is default for your enum, like false is default for bool

File: Solution1/Solution1.Tests/ConsoleApp/Services/UserCommunicateServiceTests.cs [hunk 2]
```diff
@@ -211,13 +212,13 @@ public async Task CommunicateAsync_GetBestWeatherByArrayCityNameAsync_ShowBestWe
             {
                 { true, new List<WeatherResponseDTO>
                             {
-                                new WeatherResponseDTO() { CityName = cityName, IsSuccessfulRequest = true, Temp = temp, LeadTime = leadTime },
-                                new WeatherResponseDTO() { CityName = cityName2, IsSuccessfulRequest = true, Temp = temp2, LeadTime = leadTime2 }
+                                new WeatherResponseDTO() { CityName = cityName, RequestStatus = RequestStatus.Successful, Temp = temp, LeadTime = leadTime },
```
Comment (elizavetapavliv) [inline Solution1/Solution1.Tests/ConsoleApp/Services/UserCommunicateServiceTests.cs:215]: there should be tests for cancellation as well

File: Solution1/BusinessLayer/Enum/RequestStatus.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿namespace BusinessLayer.Enum
+{
+    public enum RequestStatus
```
Comment (elizavetapavliv) [inline Solution1/BusinessLayer/Enum/RequestStatus.cs:3]: ResponseStatus

Comment (YauheniRaby) [reply Solution1/BusinessLayer/Enum/RequestStatus.cs:3]: +

File: Solution1/Solution1.Tests/ConsoleApp/Services/UserCommunicateServiceTests.cs (hunk 2 above)
Comment (YauheniRaby) [reply Solution1/Solution1.Tests/ConsoleApp/Services/UserCommunicateServiceTests.cs:215]: +

File: Solution1/BusinessLayer/Services/WeatherService.cs (hunk 1 above)
Comment (YauheniRaby) [reply Solution1/BusinessLayer/Services/WeatherService.cs:99]: +

File: Solution1/BusinessLayer/Enum/RequestStatus.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿namespace BusinessLayer.Enum
```
[the remark this answers is not in the document]
Comment (YauheniRaby) [reply Solution1/BusinessLayer/Enum/RequestStatus.cs:1]: +

File: Solution1/BusinessLayer/Command/BestWeatherCommand.cs [hunk 5]
```diff
@@ -1,15 +1,13 @@
 ﻿using BusinessLayer.Command.Abstract;
 using BusinessLayer.DTOs;
+using BusinessLayer.DTOs.Enum;
```
Comment (elizavetapavliv) [inline Solution1/BusinessLayer/Command/BestWeatherCommand.cs:3]: get used to call folder in plural: Dtos, Enums - you save in folders usually not one file

Comment (YauheniRaby) [reply Solution1/BusinessLayer/Command/BestWeatherCommand.cs:3]: +

File: Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs [hunk 6]
```diff
@@ -110,9 +109,37 @@ public async Task GetForecastByCityNameAsync_ReturnedForecastWeatherApiDTO_Succe
...
+        {
+            // Arrange
+            var cancellationTokenSource = new CancellationTokenSource();
+            cancellationTokenSource.Cancel();
+            SetHttpHandlerSettingsForTaskCanceledException();
+
+            // Act
+            
+            // Assert
+            await Assert.ThrowsAsync<TaskCanceledException>(async() => await _weatherApiService.GetByCityNameAsync(cityName, cancellationTokenSource.Token));
```
Comment (elizavetapavliv) [inline Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs:123]: FYI: its both Act / Assert in one command

File: Solution1/BusinessLayer/Services/WeatherService.cs [hunk 7]
```diff
@@ -72,38 +80,56 @@ public async Task<ForecastWeatherDTO> GetForecastByCityNameAsync(string cityName
...
             var weatherResponses = await Task.WhenAll(listTasksRequest);
             var result = weatherResponses
-                            .GroupBy(w => w.IsSuccessfulRequest)
+                            .GroupBy(w => w.ResponseStatus)
                             .ToDictionary(k => k.Key, v => v.Select(response => response));
             return result;
-        }        
+        }
+        
+        private CancellationToken GetCancellationToken()
```
Comment (elizavetapavliv) [inline Solution1/BusinessLayer/Services/WeatherService.cs:125]: Cancellation token should be generated on the top level, why in the middle?

Comment (elizavetapavliv) [reply Solution1/BusinessLayer/Services/WeatherService.cs:125]: your operation started earlier then Weather Service, and why Weather service becomes responsible for token generation?

File: Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs [hunk 8]
```diff
@@ -123,5 +150,16 @@ private void SetHttpHandlerSettings(Mock<HttpMessageHandler> httpMessageHandler,
                   ItExpr.IsAny<CancellationToken>())
                .ReturnsAsync(response);
         }
+
+        private void SetHttpHandlerSettingsForTaskCanceledException()
```
Comment (elizavetapavliv) [inline Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs:154]: strange, why TaskCanceledException thrown from httpMessageHandler?..

File: Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs [hunk 9]
```diff
@@ -168,37 +175,79 @@ public async Task GetWeatherByArrayCityNameAsync_ReturnedDictionaryForWeatherRes
...
-            Assert.Equal(3, result[false].Count());
-
-            Assert.Contains(result[true], weatherResponse => weatherResponse.CityName == cityName);
-            Assert.Contains(result[true], weatherResponse => weatherResponse.CityName == cityName2);
-            Assert.Contains(result[false], weatherResponse => weatherResponse.CityName == cityName3 
-                                                              && weatherResponse.ErrorMessage == errorMessage3); 
-            Assert.Contains(result[false], weatherResponse => weatherResponse.CityName == cityName4 
+            Assert.Contains(result.Keys, k => k == ResponseStatus.Successful);
+            Assert.Contains(result.Keys, k => k == ResponseStatus.Fail);
+            Assert.DoesNotContain(result.Keys, k => k == ResponseStatus.Canceled);
```
Comment (elizavetapavliv) [inline Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs:180]: is there any reason why for Canceled case separate test? your method can return either successful/failed, either canceled? they cannot exist together?

File: Solution1/BusinessLayer/Services/WeatherService.cs (hunk 7 above)
Comment (YauheniRaby) [reply Solution1/BusinessLayer/Services/WeatherService.cs]: +

File: Solution1/BusinessLayer/Command/BestWeatherCommand.cs [hunk 10]
```diff
@@ -1,28 +1,29 @@
...
 using System.Threading.Tasks;
 
 namespace BusinessLayer.Command
 {
-    public class BestWeatherCommand : ICommand<Dictionary<bool, IEnumerable<WeatherResponseDTO>>>
+    public class BestWeatherCommand : ICommand<Dictionary<ResponseStatus, IEnumerable<WeatherResponseDTO>>>
     {
         private readonly IWeatherServiсe _weatherServiсe;
         private readonly IEnumerable<string> _cityNames;
+        private readonly CancellationToken _cancellationToken;
```
Comment (elizavetapavliv) [inline Solution1/BusinessLayer/Command/BestWeatherCommand.cs:15]: no no no

File: Solution1/BusinessLayer/Command/BestWeatherCommand.cs [hunk 11]
```diff
@@ -1,28 +1,29 @@
...
 
-        public BestWeatherCommand(IWeatherServiсe weatherServiсe, IEnumerable<string> cityNames)
+        public BestWeatherCommand(IWeatherServiсe weatherServiсe, IEnumerable<string> cityNames, CancellationToken cancellationToken = new CancellationToken())
```
Comment (elizavetapavliv) [inline Solution1/BusinessLayer/Command/BestWeatherCommand.cs:17]: noooo

Comment (elizavetapavliv) [reply Solution1/BusinessLayer/Command/BestWeatherCommand.cs:17]: Cancelation token generated ONCE on the top level of application, and passed throw all the path

File: Solution1/ConsoleApp/Services/UserCommunicateService.cs [hunk 12]
```diff
@@ -157,39 +160,66 @@ private async Task GetBestWeatherByArrayCityNameAsync()
                 return;
             }
 
-            var command = new BestWeatherCommand(_weatherServiсe, arrayCityNames.Split(',').Select(cityName => cityName.Trim()));
+            CancellationToken cancellationToken;
```
Comment (elizavetapavliv) [inline Solution1/ConsoleApp/Services/UserCommunicateService.cs:163]: 1) use ternary operator
2) use CancellationToken.None

File: Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs [hunk 13]
```diff
@@ -45,10 +44,10 @@ public async Task GetByCityNameAsync_ReturnedWeatherApiDTO_Success()
                 Content = new StringContent(JsonSerializer.Serialize(new { Main = new { Temp = temp }, Name = cityName }, _serializerOptions)),
             };
 
-            SetHttpHandlerSettings(_httpMessageHandler, response, urlString);
+            SetHttpHandlerSettings(response, urlString);
 
             // Act
-            var result = await _weatherApiService.GetByCityNameAsync(cityName);
+            var result = await _weatherApiService.GetByCityNameAsync(cityName, new CancellationToken());
```
Comment (elizavetapavliv) [inline Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs:50]: CancellationToken.None when you don't need smth specific from it

File: Solution1/BusinessLayer/Command/BestWeatherCommand.cs (hunk 10 above)
Comment (YauheniRaby) [reply Solution1/BusinessLayer/Command/BestWeatherCommand.cs:15]: +

File: Solution1/BusinessLayer/Command/BestWeatherCommand.cs (hunk 11 above)
Comment (YauheniRaby) [reply Solution1/BusinessLayer/Command/BestWeatherCommand.cs:17]: +

File: Solution1/ConsoleApp/Services/UserCommunicateService.cs (hunk 12 above)
Comment (YauheniRaby) [reply Solution1/ConsoleApp/Services/UserCommunicateService.cs:163]: +

File: Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs (hunk 13 above)
Comment (YauheniRaby) [reply Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs:50]: +

File: Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs (hunk 9 above)
Comment (YauheniRaby) [reply Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs:180]: +

File: Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs (hunk 8 above)
Comment (YauheniRaby) [reply Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs:152]: +

File: Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs (hunk 6 above)
Comment (YauheniRaby) [reply Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs:121]: +

File: Solution1/ConsoleApp/Services/UserCommunicateService.cs [hunk 14]
```diff
@@ -158,38 +161,57 @@ private async Task GetBestWeatherByArrayCityNameAsync()
...
+                ShowDebugInformationForCanceledRequests(canceledWeatherResponses, "On canceled:");
             }
 
             return;
         }
 
         private void ShowDebugInformation(IEnumerable<WeatherResponseDTO> responses, string header, string propertyName)
         {
-            Console.WriteLine(
+            if (responses != null)
```
Comment (elizavetapavliv) [inline Solution1/ConsoleApp/Services/UserCommunicateService.cs:195]: what is the use of .? operator?

File: Solution1/ConsoleApp/Services/UserCommunicateService.cs [hunk 15]
```diff
@@ -158,38 +161,57 @@ private async Task GetBestWeatherByArrayCityNameAsync()
...
                 responses
                     ?.Aggregate(
                         $"{header}",
                         (result, next) => $"{result}{Environment.NewLine}{GetRepresentationResponse(next, propertyName)}"));
+            }
+        }
+
+        private void ShowDebugInformationForCanceledRequests(IEnumerable<WeatherResponseDTO> responses, string header)
+        {
+            if (responses != null)
```
Comment (elizavetapavliv) [inline Solution1/ConsoleApp/Services/UserCommunicateService.cs:207]: same question

File: Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs (hunk 13 above)
Comment (elizavetapavliv) [reply Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs:50]: ?

File: Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs [hunk 16]
```diff
@@ -88,11 +87,11 @@ public async Task GetForecastByCityNameAsync_ReturnedForecastWeatherApiDTO_Succe
...
             };
 
-            SetHttpHandlerSettings(_httpMessageHandler, responseCoordinates, urlCoordinates);
-            SetHttpHandlerSettings(_httpMessageHandler, responseForecast, urlForecast);      
+            SetHttpHandlerSettings(responseCoordinates, urlCoordinates);
+            SetHttpHandlerSettings(responseForecast, urlForecast);      
             
             // Act
-            var result = await _weatherApiService.GetForecastByCityNameAsync(cityName, listWeatherAnonymousObject.Count());
+            var result = await _weatherApiService.GetForecastByCityNameAsync(cityName, listWeatherAnonymousObject.Count(), new CancellationToken());
```
Comment (elizavetapavliv) [inline Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs:94]: none

File: Solution1/ConsoleApp/Services/UserCommunicateService.cs (hunk 14 above)
Comment (YauheniRaby) [reply Solution1/ConsoleApp/Services/UserCommunicateService.cs:194]: +

File: Solution1/ConsoleApp/Services/UserCommunicateService.cs (hunk 15 above)
Comment (YauheniRaby) [reply Solution1/ConsoleApp/Services/UserCommunicateService.cs:206]: +

File: Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs (hunk 16 above)
Comment (YauheniRaby) [reply Solution1/Solution1.Tests/BL/Services/WeatherApiServiceTests.cs:94]: +

File: Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs [hunk 17]
```diff
@@ -128,77 +133,87 @@ public async Task GetWeatherByArrayCityNameAsync_ReturnedDictionaryForWeatherRes
...
             var validationFailure = new List<ValidationFailure>() { new ValidationFailure("CityName", "Test validation error")};
             var notValidResult = new ValidationResult(validationFailure);
             var validResult = new ValidationResult(new List<ValidationFailure>());
 
             SetWeatherApiServiceSettings(weatherApiDto1);
             SetWeatherApiServiceSettings(weatherApiDto2);
             SetWeatherApiServiceExceptionSettings(cityName3, new Exception(errorMessage3));
-            SetWeatherApiServiceExceptionSettings(cityName4, new  ValidationException(validationFailure));
+            SetWeatherApiServiceExceptionSettings(cityName4, new ValidationException(validationFailure));
+            SetWeatherApiServiceExceptionSettings(cityName6, new TaskCanceledException());
```
[the remark this answers is not in the document]
Comment (elizavetapavliv) [reply Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs:156]: it's correct test GetForecastByCityNameAsync_GenerateOperationCanceledException_Success()

File: Solution1/BusinessLayer/Services/WeatherService.cs [hunk 18]
```diff
@@ -68,42 +72,53 @@ public async Task<ForecastWeatherDTO> GetForecastByCityNameAsync(string cityName
...
                     else
                     {
+                        weatherResponseDTO.ResponseStatus = ResponseStatus.Fail;
                         weatherResponseDTO.ErrorMessage = validationResult.Errors.FirstOrDefault().ErrorMessage;
-                    }                    
+                    }
+                }
+                catch (OperationCanceledException)
+                {
+                    weatherResponseDTO.ErrorMessage = "Timeout exceeded";
```
Comment (elizavetapavliv) [inline Solution1/BusinessLayer/Services/WeatherService.cs:103]: why it's not in Constants?...

File: Solution1/BusinessLayer/Services/WeatherService.cs [hunk 19]
```diff
@@ -68,42 +72,53 @@ public async Task<ForecastWeatherDTO> GetForecastByCityNameAsync(string cityName
...
                         if (temp.HasValue)
                         {
                             weatherResponseDTO.Temp = temp.Value;
-                            weatherResponseDTO.IsSuccessfulRequest = true;
+                            weatherResponseDTO.ResponseStatus = ResponseStatus.Successful;
                         }
                         else
                         {
+                            weatherResponseDTO.ResponseStatus = ResponseStatus.Fail;
                             weatherResponseDTO.ErrorMessage = "Unknown error";
```
Comment (elizavetapavliv) [inline Solution1/BusinessLayer/Services/WeatherService.cs:92]: why it's not in Constants?...

File: Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs [hunk 20]
```diff
@@ -52,12 +54,13 @@ public async Task GetByCityNameAsync_ReturnedWeatherDTO_Success()
...
-                    weatherApiService
-                    .GetByCityNameAsync(cityName))
+                    weatherApiService.GetByCityNameAsync
+                    (It.Is<string>(x => x == cityName),
+                     It.IsAny<CancellationToken>()))
                 .ReturnsAsync(weatherApiDto);
 
             // Act
-            var result = await _weatherService.GetByCityNameAsync(cityName);
+            var result = await _weatherService.GetByCityNameAsync(cityName, CancellationToken.None);
```
Comment (elizavetapavliv) [inline Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs:63]: where is case for cancellation? same question for other methods

File: Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs [hunk 21]
```diff
@@ -100,12 +103,14 @@ public async Task GetForecastByCityNameAsync_ReturnedForecastWeatherDTO_Success(
 
             _weatherApiServiceMock
                 .Setup(weatherApiService =>
-                    weatherApiService
-                    .GetForecastByCityNameAsync(cityName, countWeatherPoints))
+                    weatherApiService.GetForecastByCityNameAsync
+                    (It.Is<string>(x => x == cityName),
+                     It.Is<int>(x => x == countWeatherPoints),
```
Comment (elizavetapavliv) [inline Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs:108]: you can pass just countWeatherPoints

File: Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs [hunk 22]
```diff
@@ -165,40 +171,62 @@ public async Task GetWeatherByArrayCityNameAsync_ReturnedDictionaryForWeatherRes
...
+            // Arrange
+            var cityName2 = "Paris";
+            var errorMessage = "Timeout exceeded";
+            var listCityName = new List<string>() { cityName, cityName2 };
+            
+            //Act
+            var result = await _weatherService.GetWeatherByArrayCityNameAsync(listCityName, new CancellationToken(true));
+
+            // Assert
+            Assert.Contains(result.Keys, k => k == ResponseStatus.Canceled);
```
Comment (elizavetapavliv) [inline Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs:205]: what if result is null? or Keys is null?

File: Solution1/BusinessLayer/Services/WeatherService.cs (hunk 18 above)
Comment (YauheniRaby) [reply Solution1/BusinessLayer/Services/WeatherService.cs]: +

File: Solution1/BusinessLayer/Services/WeatherService.cs (hunk 19 above)
Comment (YauheniRaby) [reply Solution1/BusinessLayer/Services/WeatherService.cs]: +

File: Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs (hunk 20 above)
Comment (YauheniRaby) [reply Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs:63]: +

File: Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs (hunk 21 above)
Comment (YauheniRaby) [reply Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs:108]: +

File: Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs (hunk 22 above)
Comment (YauheniRaby) [reply Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs:205]: +

File: Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs (hunk 21 above)
Comment (elizavetapavliv) [reply Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs:112]: not resolved

File: Solution1/BusinessLayer/Constants.Errors.cs [hunk 23]
```diff
@@ -0,0 +1,11 @@
+﻿namespace BusinessLayer
+{
+    public static partial class Constants
```
Comment (elizavetapavliv) [inline Solution1/BusinessLayer/Constants.Errors.cs:3]: why other constants in Console app

Comment (YauheniRaby) [reply Solution1/BusinessLayer/Constants.Errors.cs:3]: +

File: Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (YauheniRaby) [reply Solution1/Solution1.Tests/BL/Services/WeatherServiceTests.cs]: +

**Review by elizavetapavliv — APPROVED** [review]

**Pull request merged**